crates/polars-core/src/series/any_value.rs

```diff
@@ -449,103 +451,60 @@ impl Series {
         Ok(s)
     }
 
-    pub fn from_any_values(name: &str, avs: &[AnyValue], strict: bool) -> PolarsResult<Series> {
-        let mut all_flat_null = true;
-        match avs.iter().find(|av| {
-            if !matches!(av, AnyValue::Null) {
-                all_flat_null = false;
-            }
-            !av.is_nested_null()
-        }) {
-            None => {
-                if all_flat_null {
-                    Ok(Series::new_null(name, avs.len()))
-                } else {
-                    // second pass and check for the nested null value that toggled `all_flat_null` to false
-                    // e.g. a list<null>
-                    if let Some(av) = avs.iter().find(|av| !matches!(av, AnyValue::Null)) {
-                        let dtype: DataType = av.into();
-                        Series::from_any_values_and_dtype(name, avs, &dtype, strict)
+    /// Construct a new [`Series`]` from a slice of AnyValues.
+    ///
+    /// The data type of the resulting Series is determined by the `values`
+    /// and the `strict` parameter:
+    /// - If `strict` is `true`, the data type is equal to the data type of the
+    ///   first non-null value. If any other non-null values do not match this
+    ///   data type, an error is raised.
+    /// - If `strict` is `false`, the data type is the supertype of the
+    ///   `values`. The supertype is determined by a pass over the values.
+    ///   Values encountered that do not match the supertype are set to null.
+    /// - If no values were passed, the resulting data type is `Null`.
+    pub fn from_any_values(name: &str, values: &[AnyValue], strict: bool) -> PolarsResult<Series> {
```

Comment:
This is also used in performance critical code. We should definitely not do a full pass by default. E.g. there is a lot of code that uses this where all types are equal. We should create wrapper method that does the full pass and leave this one as much single pass as possible.

Reply:
The full pass is only done when `strict=false`. Otherwise, we can just take the first non-null (like before) as we know mixed types will raise anyway.

I could find only one reference to `from_any_values` with `strict=false` in performance critical code: the `finalize` method for `SortSink`. And there it seems the dtype should be equal for all samples, so `strict` should be `true` there I believe (updated in new commit).

So unless I am missing something, we're not introducing any performance penalties here.

The way I see it, `from_any_values` _is_ the wrapper method for `from_any_values_and_dtype`, and you can use the `strict` parameter to choose which type of dtype inference you want (first non null or supertype). And in this case it's fair to assume `strict=false` should be less performant as it needs to do some work to determine fallbacks (supertype), where `strict=true` can just raise.

Reply:
Ah, I was wrong. I though we used it in `pivot` as well, but we use `from_any_value_and_dtype` there. Ok, can you make this super clear in the docstring? I don't like implicit full passes. ;)

Reply:
Done!